### PR TITLE
Introduce universal poll retries function

### PR DESCRIPTION
**Pull request by nathanleclaire**
I feel like we need something more uniform than the current situation - there are lots and lots of places where this pattern is needed, e.g. https://github.com/docker/machine/pull/665

Think it's important to have an answer so that contributors don't accidentally spin forever without a max attempts or spam a lot by forgetting a sleep.

We can move over instances of this pattern to this universal way as we see them.

Let me know what you think @ehazlett @sthulb 

Signed-off-by: Nathan LeClaire <nathan.leclaire@gmail.com>

### CODE REVIEW

**Comment by sthulb** [comment]
I like it, can you change the defaults to 10 retries over 10 seconds?

**Comment by ehazlett** [comment]
+1 -- @sthulb i like it at 1 second.  This helps keep it fast as most operations will resolve within a few seconds and gives a good default 1 min timeout IMO.

**Comment by sthulb** [comment]
I think 1 minute is too long.

**Comment by ehazlett** [comment]
It's trying every 1 second for a max of 1 minute.  You think that's too long?

**Comment by sthulb** [comment]
Yeah, I think the default should be lower, the things that we know require more can obviously just set it using the more verbose function.

**Comment by ehazlett** [comment]
Ok so the interval of 1 second is fine, you just want a default max of 10 seconds?

**Comment by sthulb** [comment]
yep!

**Comment by ehazlett** [comment]
Cool.  I think we could try that and see what happens.  @nathanleclaire ?

**Comment by nathanleclaire** [comment]
I'm OK with it.  I've just realized I should update the "wait until the instance is running" timeout to be a few minutes or so before we merge it in though, since it takes a while for newly created instances to boot up.

**Comment by ehazlett** [comment]
@nathanleclaire lol yeah that probably would be a good thing :)

File: drivers/amazonec2/amazonec2.go [hunk 1]
```diff
@@ -310,17 +310,18 @@ func (d *Driver) Create() error {
 
 	d.InstanceId = instance.InstanceId
 	log.Debug("waiting for ip address to become available")
-	for {
+	if err := utils.WaitFor(func() bool {
```
Comment (cjellick) [inline drivers/amazonec2/amazonec2.go]: I think pulling the anonymous function out of the if statement would make this code much more readable.

File: drivers/amazonec2/amazonec2.go [hunk 2]
```diff
@@ -519,29 +521,6 @@ func (d *Driver) sshKeyPath() string {
 	return path.Join(d.storePath, "id_rsa")
 }
 
-func (d *Driver) updateDriver() error {
```
Comment (nathanleclaire) [inline drivers/amazonec2/amazonec2.go]: This didn't seem to be doing anything which wasn't already done by the main `Create` function or the other check for IP address, so I've taken it out.  Instance creation still works fine, so let me know if I'm overlooking something.

**Comment by nathanleclaire** [comment]
OK I've pulled the closures out into real methods.  I'm waffling on the default timeout though; I have it set to 3 seconds interval / 3 minutes total here: @sthulb could you please elaborate on your concern?  I'd rather give a comfortable amount of space for resources to come up by default than accidentally bail when everything is fine.

File: drivers/amazonec2/amazonec2.go (hunk 2 above)
Comment (sthulb) [inline drivers/amazonec2/amazonec2.go]: The IP address on EC2 instances normally changes on stop/start, I guess this function was for that use case.

**Comment by sthulb** [comment]
@nathanleclaire I think 1 minute is too long. Perhaps 10 seconds is too short though, there's probably a middle ground somewhere :)

File: drivers/amazonec2/amazonec2.go [hunk 3]
```diff
@@ -610,6 +593,17 @@ func (d *Driver) isSwarmMaster() bool {
 	return d.SwarmMaster
 }
 
+func (d *Driver) securityGroupAvailableFunc(id string) func() bool {
```
Comment (ehazlett) [inline drivers/amazonec2/amazonec2.go]: This is a bit nit picky but do we need to name this with "Func" ?

Comment (nathanleclaire) [inline drivers/amazonec2/amazonec2.go]: Hm, I don't really care either way but I called it that way to be more clear to the reader of the code that it actually returns a function as opposed to the others which return a boolean.  Happy to take it out if you think it's confusing or redundant.

Comment (ehazlett) [inline drivers/amazonec2/amazonec2.go]: nah that works

**Comment by ehazlett** [comment]
I think this is good for now.  Until we get some more feedback I think the 3/3 is a good start to test with.  We can always update it if we find it's not enough or too much.

**Comment by ehazlett** [comment]
LGTM

**Pull request merged**